whole-tale/ngx-dashboard
# Allow user to upload entire folder structures

## Problem
It would be nice to be able to upload an entire folder at once. This is currently possible by zipping up the folder, uploading the zip, starting up a Tale, and using the Tale's console/terminal to unzip the folder via CLI. This is obviously a bit cumbersome for standard users, so we'd like to add a convenience method to easily enable people to upload an entire folder while maintaining its underlying structure

Fixes #233 

## Approach
Uses https://github.com/whole-tale/virtual_resources/pull/20

* Added new "folder upload" events to `file-browser` / `tale-files`
* Added a new file input with `directory` attributes: `<input type="file" directory webkit directory mozdirectory ... />`
* Added new `uploadFolder` method to `tale-files` - this ties in with our existing `uploadFiles`

## How to Test
Prerequisites: at least one Tale created, run with https://github.com/whole-tale/virtual_resources/pull/20

1. Checkout this branch locally, rebuild the dashboard
2. Login to the WholeTale Dashboard
3. Create a Tale, if you haven't already
4. Navigate to Run > Files > Home for a Tale that you own
5. At the top-right, click on the plus (+) button
    * You should see a dropdown appear
    * You should see a new dropdown option: "Upload Folder..."
6. Click the "Upload Folder..." option
    * You should see a folder picker dialog pop up
    * You should notice that this folder picker dialog only allows you to select a folder
7. Choose a folder and click "Upload" at the bottom-right of the dialog
    * You should see the dialog close, and your uploads should begin
    * You should see the folder you selected now appears in the Files menu
8. Click down to navigate into the folder you uploaded
    * You should see all of the files from the folder that you chose are uploaded/uploading here

* Repeat steps for Run > Files > Tale Workspace

## TODOs
* ~~annoying scoping bug within `uploadFolder(filesToUpload)` where `filesToUpload` is appearing empty in rxjs callbacks~~
* ~~EXACERBATED ISSUE: multiple uploads bog down the backend, so we may need to throttle the number of active uploads in the queue~~ uploads now happen one at a time
* ~~KNOWN ISSUE: in-progress uploads do not properly render a progress bar or updates after folder navigation (move upload queue to notifications? pop up file queue?)~~ in-progress uploads should now show progress, even after navigating to subfolders
* KNOWN ISSUE: a hard refresh of the page will disrupt any in-progress downloads (not sure that we can get around this)
* ~~drag+drop functionality is still missing, if that is still desirable~~
